src/wallet/wallet2.cpp

```diff
@@ -3404,6 +3594,209 @@ bool wallet2::load_tx(const std::string &signed_filename, std::vector<tools::wal
+  for (auto &ptx: exported_txs.m_ptx) LOG_PRINT_L0(cryptonote::obj_to_json_str(ptx.tx));
+
+  if (accept_func && !accept_func(exported_txs))
+  {
+    LOG_PRINT_L1("Transactions rejected by callback");
+    return false;
+  }
+
+  const bool is_signed = exported_txs.m_signers.size() >= m_multisig_threshold;
+  if (is_signed)
+  {
+    for (const auto &ptx: exported_txs.m_ptx)
+    {
+      const crypto::hash txid = get_transaction_hash(ptx.tx);
+      if (store_tx_info())
+      {
+        m_tx_keys.insert(std::make_pair(txid, ptx.tx_key));
+      }
+    }
+  }
+
+  return true;
+}
+//----------------------------------------------------------------------------------------------------
+bool wallet2::sign_multisig_tx(multisig_tx_set &exported_txs, const std::string &filename, std::vector<crypto::hash> &txids)
+{
+  THROW_WALLET_EXCEPTION_IF(exported_txs.m_ptx.empty(), error::wallet_internal_error, "No tx found");
+
+  // sign the transactions
+  for (size_t n = 0; n < exported_txs.m_ptx.size(); ++n)
+  {
+    tools::wallet2::pending_tx &ptx = exported_txs.m_ptx[n];
+    tools::wallet2::tx_construction_data &sd = ptx.construction_data;
+    LOG_PRINT_L1(" " << (n+1) << ": " << sd.sources.size() << " inputs, mixin " << (sd.sources[0].outputs.size()-1) <<
+        ", signed by " << exported_txs.m_signers.size() << "/" << m_multisig_threshold);
+    cryptonote::transaction tx;
+    std::vector<rct::multisig_out> msout;
+    bool r = cryptonote::construct_tx_with_tx_key(m_account.get_keys(), sd.sources, sd.splitted_dsts, sd.extra, tx, sd.unlock_time, ptx.tx_key, sd.use_rct, &msout);
+    THROW_WALLET_EXCEPTION_IF(!r, error::tx_not_constructed, sd.sources, sd.splitted_dsts, sd.unlock_time, m_testnet);
+
+    // since this is multisig, our local secret spend key will generate wrong key images,
+    // so we copy the incoming key images to the generated tx since this is expected
+    THROW_WALLET_EXCEPTION_IF(tx.vin.size() != ptx.tx.vin.size(),
+        error::wallet_internal_error, "Transaction prefix does not match data");
+    for (size_t i = 0; i < tx.vin.size(); ++i)
+    {
+      CHECKED_GET_SPECIFIC_VARIANT(tx.vin[i], txin_to_key, in, false);
+      CHECKED_GET_SPECIFIC_VARIANT(ptx.tx.vin[i], const txin_to_key, in_org, false);
+      in.k_image = in_org.k_image;
+    }
+
+    THROW_WALLET_EXCEPTION_IF(get_transaction_prefix_hash (tx) != get_transaction_prefix_hash(ptx.tx),
+        error::wallet_internal_error, "Transaction prefix does not match data");
+
+    // Tests passed, sign
+    rct::keyV k;
+    for (size_t idx: sd.selected_transfers)
+      k.push_back(get_multisig_k(idx));
+    std::vector<unsigned int> indices;
+    for (const auto &source: sd.sources)
+      indices.push_back(source.real_output);
+    THROW_WALLET_EXCEPTION_IF(!rct::signMultisig(ptx.tx.rct_signatures, indices, k, ptx.msout, rct::sk2rct(get_account().get_keys().m_spend_secret_key)),
+        error::wallet_internal_error, "Failed signing, transaction likely malformed");
```

Comment:
I think `multisig_k` should be cleared here in order to prevent an attacker from stealing the spend secret key in the following way:

Consider a scenario with a 3/3 multisig wallet, where Alice, Bob and Charlie are the participants. Alice and Charlie are colluding together to steal the spend secret key from Bob. The shared view secret key is `a` while the spend secret keys of Alice, Bob and Charlie are `b_Alice`, `b_Bob` and `b_Charlie`, respectively.

### 1. Alice creates a transaction and passes it to Bob.

At this point, the partial signature scalar amounts to:

```
s_step1 = k_Alice - c * (Hs(a*R) + b_Alice)
```

### 2. Bob signs it with his spend secret key and passes it to Charlie.

At this point, the partial signature scalar amounts to:

```
s_step2 = k_Alice + k_Bob - c * (Hs(a*R) + b_Alice + b_Bob)
```

### 3. Charlie shares the partially signed transaction with Alice.

At this point, Alice and Charlie can know Bob's contribution to the signature scalar by taking the difference:

```
s_diff = s_step2 - s_step1
       = k_Bob - c * b_Bob
```

### 4. Alice creates another transaction and passes it to Bob, saying "Sorry, there was a mistake in the transaction I made previously, so please sign again this new transaction."

### 5. Bob signs the new transaction and passes it to Charlie.

### 6. Charlie shares the partially signed new transaction with Alice.

At this point, Alice and Charlie do exactly the same computation as above, and obtain another instance of Bob's contribution to the signature scalar using his same `k_Bob` and a different `c_new`:

```
s_diff_new = k_Bob - c_new * b_Bob
```

### 7. Alice and Charlie can extract Bob's spend secret key as
```
b_Bob = (s_diff_new - s_diff) / (c_new - c)
```

To prevent this attack, the `multisig_k` should never be reused, and it should be cleared every time a participant signs a transaction (either via `transfer` or `sign_multisig`). If a partially signed transaction is to be cancelled, the export/import of multisig info should be performed again.